pkg/controller/apiextensions/composite/api.go

```diff
@@ -125,6 +151,93 @@ func (r *APISelectorResolver) ResolveSelector(ctx context.Context, cp resource.C
 	return errors.New(errNoCompatibleComposition)
 }
 
+// NewAPIDefaultCompositionSelector returns a APIDefaultCompositionSelector.
+func NewAPIDefaultCompositionSelector(c client.Client, ref v1.ObjectReference) *APIDefaultCompositionSelector {
+	return &APIDefaultCompositionSelector{client: c, defRef: ref}
+}
+
+// APIDefaultCompositionSelector selects the default composition referenced in
+// the definition of the resource if neither a reference nor selector is given
+// in composite resource.
+type APIDefaultCompositionSelector struct {
+	client client.Client
+	defRef v1.ObjectReference
+}
+
+// SelectComposition selects the default compositionif neither a reference nor
+// selector is given in composite resource.
+func (r *APIDefaultCompositionSelector) SelectComposition(ctx context.Context, cp resource.Composite) error {
+	if cp.GetCompositionReference() != nil || cp.GetCompositionSelector() != nil {
+		return nil
+	}
+	def := &v1alpha1.InfrastructureDefinition{}
+	if err := r.client.Get(ctx, meta.NamespacedNameOf(&r.defRef), def); err != nil {
+		return errors.Wrap(err, errGetInfraDef)
+	}
+	if def.Spec.DefaultCompositionRef == nil {
+		return nil
+	}
+
+	// TODO(muvaf): A validating webhook on InfrastructureDefinition that will make
+	// sure referenced Composition is compatible could make more sense.
```

Comment:
I agree a webhook would be the better approach. Do we need to check this in this controller? What happens today if a composite explicitly references a composition that is not compatible with it?

Reply:
> Do we need to check this in this controller?

Actually, InfraDef controller makes more sense since the change happens there, not on composite resource.

> What happens today if a composite explicitly references a composition that is not compatible with it?

Nothing, really. The compatibility is merely a documentation plus the guarantee in the selection, which got me thinking maybe it should be a lower level guarantee, i.e. check the reference in main `Reconcile` for compatibility. Though we'd still need the same check in label selector because it fetches a list of `Composition` and then do the filtering.

I think what I'll do is to make sure infradef controller returns error if an incompatible default or enforced composition is given, don't check compatibility in these two composition selectors, check the compatibility of the actual reference in `Reconcile` of composite resource. That'd fix the bug where the given direct reference is incompatible.

Reply:
I ended up implementing the check only for the composition reference. While testing, I realized that InfraDef doing a check is not really meaningful because it doesn't block the usage of that reference on composite level. In fact, I think it shouldn't even block CRD rendering, for example, even though default composition is not compatible, that default comp ref may never be used. So, it just seemed like not very meaningful check. If it's wrong you'll see error on composite which is the instance where this error should actually block. It could still be nice to have it but I'm not sure if additional logic for fetching composition at every reconcile would be worth the benefit.

I also added events when enforced or default composition has been selected so that people don't go crazy about where that composition reference came from.

Also, I removed the TODO there, because if definition and referenced composition is created at the same time with `kubectl apply`, then validating webhook might reject it since the referenced composition is yet to be created.

Reply:
Making it `+immutable` comes with a caveat actually. If I create infradef with an enforced composition ref and then I might want to remove that enforcement. I can't do that right now because the controller would have to restarted; even if it checks InfraDef at every reconcile, `+immutable` would lead the same behavior. Are we OK with `once enforced, you need to delete infradef to not enforce it` approach? When I manually tested, I did `kubectl apply` with infradef that doesn't have enforcement anymore and hit this. But I'm not sure if that'd be a common thing.

Reply:
> Are we OK with once enforced, you need to delete infradef to not enforce it approach?

I feel okay with this limitation.